design/fake-client.rst

```diff
@@ -0,0 +1,153 @@
+zhmcclient.
+
+Requirements
+------------
+
+* Must avoid the need for a real HMC and still allow usage of the zhmcclient
+  API.
+* Both read-only and read-write operations must be supported.
+* Read-write operations must reflect their changes properly (i.e. they must be
+  visible in corresponding read operations).
+* The resource state of the faked HMC must be preloaded (e.g. for resources
+  that cannot be created by clients), and its resource descriptions must be
+  easy to use (i.e. not in terms of REST URIs.
+* It must be possible to define testcases with error injection.
+
+It seems that with these requirements, we can assume that there is a need for
+a faked HMC that can represent its resources and properties as Python objects,
+so that changes can be applied consistently.
+
+Design options
+--------------
+
+There are two basic design options for the fake client:
+
+* Complete re-implementation of the zhmcclient API, that behaves the same
+  way as the zhmcclient package, but consists of a different classes that
+  implement a faked HMC.
+* Usage of the real zhmcclient package, and replacing its Session object
+  with a fake session that implements a faked HMC.
+
+The first design option is probably less effort, but it has the drawback that
+subtle changes in the API may be forgotten. One example for that is the
+recent introduction of an optimization for findall() by name, which now uses
+a cached mapping of names to URIs, but had the subtle externally visible
+change that the resource objects returned by findall(name) now only have
+a very minimal set of properties, and no longer those that are returned by
+list() which was the earlier implementation. For example, the "status"
+property is no longer set in that case. The second design would automatically
+represent this example of a change.
+
+Design
+------
+
+In the following description, the second design option is used.
+
+There is a fake session object that the user of the zhmcclient package
+will use instead of zhmcclient.Session. The fake session object contains a
+faked HMC that can represent all resources relevant for the zhmcclient.
+There fake session object implements the same API as zhmcclient.Session, so
+that the rest of the zhmcclient classes does not need to be changed.
+The get(), post() and delete() methods of the fake session class will
+be redirected to operate on the faked HMC.
+
+The fake session object will have the ability to be "loaded" with an
+initial state of resources. The resource definitions can be provided as
+a dictionary of nested dictionaries, following this example::
+
+    resources = {
+        'cpcs': [  # name of manager attribute for this resource
+            {
+                'properties': {
+                    # object-id and object-uri are auto-generated
+                    'name': 'cpc_1',
+                    . . .  # more properties
+                },
+                'adapters': [
```

Comment:
I prefer the dict approach - as I can easily see how my hmc looks like at first sight. with the programmed approach, it's hard to see and understand the nesting at first sight. I'm fine if both are available..